### DISCUSSION TITLE
New Record types do not allow overriding the equals method

### DISCUSSION THREAD

**Discussion by tmccowan** [dotnet/csharplang#4376] [Language Ideas] [answered]
I started testing out record types in an application that I maintain and ran into some issues. The default == logic works just fine until you have a reference type property. 

Here is an example of my problem

`Person p1 = new() { FirstName = "Travis", LastName = "McCowan", Data = new byte[] { 1, 3, 5, 7, 9 } };`
`Person p2 = p1 with { Data = new byte[] { 1, 3, 5, 7, 9 } };`

`Console.WriteLine(p1 == p2); // False`

`public record Person`
`{`
`        public string FirstName { get; init; }`
`        public string LastName { get; init; }`
`        public byte[] Data { get; init; }`
`}`

Possible solutions to this issue
- Allow overriding of Equals or GetHashCode  to determine equality.
- As record types compare the value of their properties by default, it could just evaluate the reference types the same way.

`Person p1 = new() { FirstName = "Travis", LastName = "McCowan", Data = new byte[] { 1, 3, 5, 7, 9 } };`
`Person p2 = p1 with { Data = new byte[] { 1, 3, 5, 7, 9 } };`

`Console.WriteLine(p1 == p2);  //True`

`public record Person`
`{`
    `public string FirstName { get; init; }`
    `public string LastName { get; init; }`
    `public byte[] Data { get; init; }`

`    public override int GetHashCode()`
`    {`
`        unchecked`
`        {`
`            const int HashingBase = (int)2166136261;`
`            const int HashingMultiplier = 16777619;`

`            int hash = HashingBase;`
`            hash = (hash * HashingMultiplier) ^ FirstName.GetHashCode();`
`            hash = (hash * HashingMultiplier) ^ LastName.GetHashCode();`
`            hash = (hash * HashingMultiplier) ^ new BigInteger(Data).GetHashCode();`
`            return hash;`
`        }`
`    }`
`}`

**Comment by ufcpp** [comment, marked as answer]
```cs
public record Person
{
    public string FirstName { get; init; }
    public string LastName { get; init; }
    public byte[] Data { get; init; }

    // you can customize the equality by
    public virtual bool Equals(Person other)
    {
        return other != null
            && FirstName == other.FirstName
            && LastName == other.LastName
            && new BigInteger(Data) == new BigInteger(other.Data);
    }

    public override int GetHashCode()
    {
        unchecked
        {
            const int HashingBase = (int)2166136261;
            const int HashingMultiplier = 16777619;

            int hash = HashingBase;
            hash = (hash * HashingMultiplier) ^ FirstName.GetHashCode();
            hash = (hash * HashingMultiplier) ^ LastName.GetHashCode();
            hash = (hash * HashingMultiplier) ^ new BigInteger(Data).GetHashCode();
            return hash;
        }
    }
}
```

**Reply by tmccowan** [reply]
@ufcpp Thank you for the sample. I'll use your suggestion in my project so I can move forward with record types. 
It would still be nice for record types to work out of the box for all types.
